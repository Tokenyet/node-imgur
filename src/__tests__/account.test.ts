jest.mock('node-fetch', () => jest.fn(() => Promise.resolve()));
const fetch = require.requireMock('node-fetch');

import {
  generateAccessToken,
  getBaseInfo,
  getBlockStatus,
  getBlocks,
  createBlock,
  deleteBlock,
  getImages,
  getGalleryFavorites,
  getFavorites,
  OAUTH2_TOKEN_ENDPOINT,
  ACCOUNT_ENDPOINT,
  BLOCK_STATUS_ENDPOINT,
  BLOCKS_ENDPOINT,
  BLOCK_CREATE_DELETE_ENDPOINT,
  IMAGES_ENDPOINT,
  GALLERY_FAVORITES_ENDPOINT,
  FAVORITES_ENDPOINT,
} from '../account';

import { URLSearchParams } from 'url';

afterEach(() => {
  fetch.mockClear();
});

describe('getAccessToken tests', () => {
  test('getAccessToken returns a resolved Promise on success', async () => {
    const credentials = {
      refreshToken: 'myRefreshToken',
      clientId: 'myId',
      clientSecret: 'mySecret',
    };

    await expect(generateAccessToken(credentials)).resolves.toBe(undefined);
  });

  test('getAccessToken calls the correct endpoint', async () => {
    const credentials = {
      refreshToken: 'myRefreshToken',
      clientId: 'myId',
      clientSecret: 'mySecret',
    };

    await generateAccessToken(credentials);

    const expectedParams = new URLSearchParams();
    expectedParams.append('refresh_token', credentials.refreshToken);
    expectedParams.append('client_id', credentials.clientId);
    expectedParams.append('client_secret', credentials.clientSecret);
    expectedParams.append('grant_type', 'refresh_token');

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(OAUTH2_TOKEN_ENDPOINT, {
      body: expectedParams,
      method: 'POST',
    });
  });
});

describe('getBaseInfo tests', () => {
  test('base returns a resolved Promise on success', async () => {
    const username = 'myUsername';
    const clientId = 'myClientId';

    await expect(getBaseInfo({ username, clientId })).resolves.toBe(undefined);
  });

  test('getBaseInfo calls the correct endpoint', async () => {
    const username = 'myUsername';
    const clientId = 'myClientId';

    await getBaseInfo({ username, clientId });

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(`${ACCOUNT_ENDPOINT}/${username}`, {
      headers: {
        Authorization: `Client-ID ${clientId}`,
      },
      method: 'GET',
    });
  });

  test('getBaseInfo calls the correct endpoint when using account_id', async () => {
    const accountId = '1234';
    const clientId = 'myClientId';

    await getBaseInfo({ clientId, accountId });

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(`${ACCOUNT_ENDPOINT}/?account_id=${accountId}`, {
      headers: {
        Authorization: `Client-ID ${clientId}`,
      },
      method: 'GET',
    });
  });
});

describe('getBlockStatus tests', () => {
  test('getBlockStatus returns a resolved Promise on success', async () => {
    const username = 'myUsername';
    const accessToken = 'myAccessToken';

    await expect(getBlockStatus({ username, accessToken })).resolves.toBe(undefined);
  });

  test('getBlockStatus calls the correct endpoint', async () => {
    const username = 'myUsername';
    const accessToken = 'myAccessToken';

    await getBlockStatus({ username, accessToken });

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(`${BLOCK_STATUS_ENDPOINT.replace('<username>', username)}`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      method: 'GET',
    });
  });
});

describe('getBlocks tests', () => {
  test('getBlocks returns a resolved Promise on success', async () => {
    const accessToken = 'myAccessToken';

    await expect(getBlocks(accessToken)).resolves.toBe(undefined);
  });

  test('getBlocks calls the correct endpoint', async () => {
    const accessToken = 'myAccessToken';

    await getBlocks(accessToken);

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(BLOCKS_ENDPOINT, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      method: 'GET',
    });
  });
});

describe('createBlock tests', () => {
  test('createBlock returns a resolved Promise on success', async () => {
    const username = 'myUsername';
    const accessToken = 'myAccessToken';

    await expect(createBlock({ username, accessToken })).resolves.toBe(undefined);
  });

  test('createBlock calls the correct endpoint', async () => {
    const username = 'myUsername';
    const accessToken = 'myAccessToken';

    await createBlock({ username, accessToken });

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(
      BLOCK_CREATE_DELETE_ENDPOINT.replace('<username>', username),
      {
        headers: {
          Authorization: `Bearer ${accessToken}`,
        },
        method: 'POST',
      },
    );
  });
});

describe('deleteBlock tests', () => {
  test('deleteBlock returns a resolved Promise on success', async () => {
    const username = 'myUsername';
    const accessToken = 'myAccessToken';

    await expect(deleteBlock({ username, accessToken })).resolves.toBe(undefined);
  });

  test('deleteBlock calls the correct endpoint', async () => {
    const username = 'myUsername';
    const accessToken = 'myAccessToken';

    await deleteBlock({ username, accessToken });

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(
      BLOCK_CREATE_DELETE_ENDPOINT.replace('<username>', username),
      {
        headers: {
          Authorization: `Bearer ${accessToken}`,
        },
        method: 'DELETE',
      },
    );
  });
});

describe('getImages tests', () => {
  test('getImages returns a resolved Promise on success', async () => {
    const accessToken = 'myAccessToken';

    await expect(getImages({ accessToken })).resolves.toBe(undefined);
  });

  test('getImages calls the correct endpoint using defaults', async () => {
    const accessToken = 'myAccessToken';

    await getImages({ accessToken });

    const expectedEndpoint = IMAGES_ENDPOINT.replace('<username>', 'me');
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(expectedEndpoint, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      method: 'GET',
    });
  });

  test('getImages calls the correct endpoint with username specified', async () => {
    const accessToken = 'myAccessToken';
    const username = 'testuser';

    await getImages({ accessToken, username });

    const expectedEndpoint = IMAGES_ENDPOINT.replace('<username>', username);
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(expectedEndpoint, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      method: 'GET',
    });
  });

  test('getGalleryFavorites calls the correct endpoint with defaults', async () => {
    const username = 'myUsername';
    const clientId = 'myClientId';

    await getGalleryFavorites({ username, clientId });

    const expectedEndpoint = GALLERY_FAVORITES_ENDPOINT.replace('<username>', username);
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(expectedEndpoint, {
      headers: {
        Authorization: `Client-ID ${clientId}`,
      },
      method: 'GET',
    });
  });

  test('getGalleryFavorites calls the correct endpoint with options', async () => {
    const username = 'myUsername';
    const clientId = 'myClientId';
    const page = 2;
    const favoriteSort = 'oldest';

    await getGalleryFavorites({ username, clientId, page, favoriteSort });

    const expectedEndpoint = `${GALLERY_FAVORITES_ENDPOINT.replace(
      '<username>',
      username,
    )}/${page}/${favoriteSort}`;
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(expectedEndpoint, {
      headers: {
        Authorization: `Client-ID ${clientId}`,
      },
      method: 'GET',
    });
  });

  test('getFavorites calls the correct endpoint with defaults', async () => {
    const username = 'myUsername';
    const accessToken = 'myAccessToken';

    await getFavorites({ username, accessToken });

    const expectedEndpoint = FAVORITES_ENDPOINT.replace('<username>', username);
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(expectedEndpoint, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      method: 'GET',
    });
  });

  test('getFavorites calls the correct endpoint with options', async () => {
    const username = 'myUsername';
    const accessToken = 'myaccessToken';
    const page = 2;
    const sort = 'oldest';

    await getFavorites({ username, accessToken, page, sort });

    const expectedEndpoint = `${FAVORITES_ENDPOINT.replace(
      '<username>',
      username,
    )}/${page}/${sort}`;
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(expectedEndpoint, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      method: 'GET',
    });
  });
});
