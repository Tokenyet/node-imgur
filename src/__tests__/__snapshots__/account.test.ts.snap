// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBlock tests createBlock calls the correct endpoint 1`] = `
Object {
  "data": Object {
    "blocked": true,
  },
}
`;

exports[`deleteBlock tests deleteBlock calls the correct endpoint 1`] = `
Object {
  "data": Object {
    "blocked": false,
  },
}
`;

exports[`getAccessToken tests getAccessToken calls the correct endpoint and resolves 1`] = `
Object {
  "access_token": "b6fb3801f9caf9968c369532ce26666d868f54fbfake",
  "account_id": 123456,
  "account_username": "testuser",
  "expires_in": 315360000,
  "refresh_token": "183b1z001c5a4be719083e5221d8f2718b184694fake",
  "scope": "",
  "token_type": "bearer",
}
`;

exports[`getBaseInfo tests getBaseInfo calls the correct endpoint and resolves 1`] = `
Object {
  "data": Object {
    "avatar": "https://imgur.com/user/testuser/avatar?maxwidth=290",
    "avatar_name": "default/T",
    "bio": null,
    "cover": "https://imgur.com/user/testuser/cover?maxwidth=2560",
    "cover_name": "default/1-space",
    "created": 1310522831,
    "id": 123456,
    "is_blocked": false,
    "pro_expiration": false,
    "reputation": 0,
    "reputation_name": "Neutral",
    "url": "testuser",
    "user_follow": Object {
      "status": false,
    },
  },
  "status": 200,
  "success": true,
}
`;

exports[`getBaseInfo tests getBaseInfo calls the correct endpoint and resolves when using account_id 1`] = `
Object {
  "data": Object {
    "avatar": "https://imgur.com/user/testuser/avatar?maxwidth=290",
    "avatar_name": "default/T",
    "bio": null,
    "cover": "https://imgur.com/user/testuser/cover?maxwidth=2560",
    "cover_name": "default/1-space",
    "created": 1310522831,
    "id": 123456,
    "is_blocked": false,
    "pro_expiration": false,
    "reputation": 0,
    "reputation_name": "Neutral",
    "url": "testuser",
    "user_follow": Object {
      "status": false,
    },
  },
  "status": 200,
  "success": true,
}
`;

exports[`getBlockStatus tests getBlockStatus calls the correct endpoint and resolves 1`] = `
Object {
  "data": Object {
    "blocked": false,
  },
}
`;

exports[`getBlocks tests getBlocks calls the correct endpoint and resolves 1`] = `
Object {
  "data": Object {
    "items": Array [],
    "next": null,
  },
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getAvailableAvatars calls the correct endpoint when authed 1`] = `
Object {
  "data": Object {
    "available_avatars": Array [
      Object {
        "location": "https://i.imgur.com/QmC4NzH_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/3D",
      },
      Object {
        "location": "https://i.imgur.com/f8wFE2Q_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/avocado",
      },
      Object {
        "location": "https://i.imgur.com/5IB1Gxc_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/bear",
      },
      Object {
        "location": "https://i.imgur.com/qnrO2Ed_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/beer",
      },
      Object {
        "location": "https://i.imgur.com/cscegSd_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/bolt",
      },
      Object {
        "location": "https://i.imgur.com/WM6zUJu_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/burger",
      },
    ],
    "available_avatars_count": 6,
    "avatars_are_default": false,
  },
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getAvailableAvatars calls the correct endpoint when unauthed 1`] = `
Object {
  "data": Object {
    "available_avatars": Array [
      Object {
        "location": "https://i.imgur.com/QmC4NzH_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/3D",
      },
      Object {
        "location": "https://i.imgur.com/f8wFE2Q_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/avocado",
      },
      Object {
        "location": "https://i.imgur.com/5IB1Gxc_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/bear",
      },
      Object {
        "location": "https://i.imgur.com/qnrO2Ed_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/beer",
      },
      Object {
        "location": "https://i.imgur.com/cscegSd_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/bolt",
      },
      Object {
        "location": "https://i.imgur.com/WM6zUJu_d.png?maxwidth=290&fidelity=grand",
        "name": "flavor/burger",
      },
    ],
    "available_avatars_count": 6,
    "avatars_are_default": false,
  },
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getFavorites calls the correct endpoint with defaults 1`] = `
Object {
  "data": Array [],
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getFavorites calls the correct endpoint with options 1`] = `
Object {
  "data": Array [],
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getGalleryFavorites calls the correct endpoint with defaults 1`] = `
Object {
  "data": Array [],
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getGalleryFavorites calls the correct endpoint with options 1`] = `
Object {
  "data": Array [],
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getImages calls the correct endpoint using defaults and resolves 1`] = `
Object {
  "data": Array [
    Object {
      "account_id": 123456,
      "account_url": "testuser",
      "ad_type": 0,
      "ad_url": "",
      "animated": true,
      "bandwidth": 0,
      "datetime": 1529788737,
      "deletehash": "abcdefg",
      "description": null,
      "favorite": false,
      "gifv": "https://i.imgur.com/qsRxU1f.gifv",
      "has_sound": false,
      "height": 497,
      "hls": "https://i.imgur.com/qsRxU1f.m3u8",
      "id": "qsRxU1f",
      "in_gallery": false,
      "in_most_viral": false,
      "is_ad": false,
      "link": "https://i.imgur.com/qsRxU1f.gif",
      "looping": true,
      "mp4": "https://i.imgur.com/qsRxU1f.mp4",
      "mp4_size": 0,
      "name": null,
      "nsfw": null,
      "section": null,
      "size": 852367,
      "tags": Array [],
      "title": null,
      "type": "image/gif",
      "views": 0,
      "vote": null,
      "width": 350,
    },
    Object {
      "account_id": 123456,
      "account_url": "testuser",
      "ad_type": 0,
      "ad_url": "",
      "animated": true,
      "bandwidth": 0,
      "datetime": 1529786616,
      "deletehash": "abcdefg",
      "description": null,
      "favorite": false,
      "gifv": "https://i.imgur.com/qEtPGdm.gifv",
      "has_sound": false,
      "height": 497,
      "hls": "https://i.imgur.com/qEtPGdm.m3u8",
      "id": "qEtPGdm",
      "in_gallery": false,
      "in_most_viral": false,
      "is_ad": false,
      "link": "https://i.imgur.com/qEtPGdm.gif",
      "looping": true,
      "mp4": "https://i.imgur.com/qEtPGdm.mp4",
      "mp4_size": 0,
      "name": null,
      "nsfw": null,
      "section": null,
      "size": 852367,
      "tags": Array [],
      "title": null,
      "type": "image/gif",
      "views": 0,
      "vote": null,
      "width": 350,
    },
  ],
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getImages calls the correct endpoint with username specified and resolves 1`] = `
Object {
  "data": Array [
    Object {
      "account_id": 123456,
      "account_url": "testuser",
      "ad_type": 0,
      "ad_url": "",
      "animated": true,
      "bandwidth": 0,
      "datetime": 1529788737,
      "deletehash": "abcdefg",
      "description": null,
      "favorite": false,
      "gifv": "https://i.imgur.com/qsRxU1f.gifv",
      "has_sound": false,
      "height": 497,
      "hls": "https://i.imgur.com/qsRxU1f.m3u8",
      "id": "qsRxU1f",
      "in_gallery": false,
      "in_most_viral": false,
      "is_ad": false,
      "link": "https://i.imgur.com/qsRxU1f.gif",
      "looping": true,
      "mp4": "https://i.imgur.com/qsRxU1f.mp4",
      "mp4_size": 0,
      "name": null,
      "nsfw": null,
      "section": null,
      "size": 852367,
      "tags": Array [],
      "title": null,
      "type": "image/gif",
      "views": 0,
      "vote": null,
      "width": 350,
    },
    Object {
      "account_id": 123456,
      "account_url": "testuser",
      "ad_type": 0,
      "ad_url": "",
      "animated": true,
      "bandwidth": 0,
      "datetime": 1529786616,
      "deletehash": "abcdefg",
      "description": null,
      "favorite": false,
      "gifv": "https://i.imgur.com/qEtPGdm.gifv",
      "has_sound": false,
      "height": 497,
      "hls": "https://i.imgur.com/qEtPGdm.m3u8",
      "id": "qEtPGdm",
      "in_gallery": false,
      "in_most_viral": false,
      "is_ad": false,
      "link": "https://i.imgur.com/qEtPGdm.gif",
      "looping": true,
      "mp4": "https://i.imgur.com/qEtPGdm.mp4",
      "mp4_size": 0,
      "name": null,
      "nsfw": null,
      "section": null,
      "size": 852367,
      "tags": Array [],
      "title": null,
      "type": "image/gif",
      "views": 0,
      "vote": null,
      "width": 350,
    },
  ],
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getSubmissions calls the correct endpoint with defaults 1`] = `
Object {
  "data": Array [],
  "status": 200,
  "success": true,
}
`;

exports[`getImages tests getSubmissions calls the correct endpoint with options 1`] = `
Object {
  "data": Array [],
  "status": 200,
  "success": true,
}
`;
